### PR TITLE
invoke Spark catalog plugin API to create tables

### PR DESCRIPTION
**Pull request by cloud-fan**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Today, `DeltaCatalog` takes care of the Spark CREATE TABLE command and calls `CreateDeltaTableCommand#run` at the end. Within this command, `spark.sessionState.catalog.createTable` is called, which bypasses any custom catalog that overwrites `spark_catalog`. `DeltaCatalog` always creates tables in the Hive Metastore.

This PR fixes this issue by calling the Spark catalog plugin API to create the table, to respect custom catalog that overwrites `spark_catalog`, such as Unity Catalog.

## How was this patch tested?

Locally tested it with Spark + Unity Catalog.

## Does this PR introduce _any_ user-facing changes?

no

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @tdas

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 1]
```diff
@@ -70,6 +70,12 @@ class DeltaCatalog extends DelegatingCatalogExtension
 
   val spark = SparkSession.active
 
+  private lazy val isUnityCatalog: Boolean = {
+    val delegateField = classOf[DelegatingCatalogExtension].getDeclaredField("delegate")
+    delegateField.setAccessible(true)
+    delegateField.get(this).getClass.getCanonicalName.startsWith("io.unitycatalog.connectors")
```
Comment (tdas) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:76]: lets make this just `io.unitycatalog`. I feel like we should rename the package to `io.unitycatalog.spark` ... connectors.spark makes it unnecessarily too long.

Comment (cloud-fan) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:76]:
```suggestion
    delegateField.get(this).getClass.getCanonicalName.startsWith("io.unitycatalog")
```

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala [hunk 2]
```diff
@@ -67,7 +67,8 @@ case class CreateDeltaTableCommand(
     operation: TableCreationModes.CreationMode = TableCreationModes.Create,
     tableByPath: Boolean = false,
     override val output: Seq[Attribute] = Nil,
-    protocol: Option[Protocol] = None)
+    protocol: Option[Protocol] = None,
+    createTableFunc: Option[CatalogTable => Unit] = None)
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala:71]: Can we update the class doc with the new parameter information?

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala [hunk 3]
```diff
@@ -631,10 +632,14 @@ case class CreateDeltaTableCommand(
...
       case _ if tableByPath => // do nothing with the metastore if this is by path
       case TableCreationModes.Create =>
-        spark.sessionState.catalog.createTable(
-          cleaned,
-          ignoreIfExists = existingTableOpt.isDefined || mode == SaveMode.Ignore,
-          validateLocation = false)
+        if (createTableFunc.isDefined) {
+          createTableFunc.get.apply(cleaned)
+        } else {
+          spark.sessionState.catalog.createTable(
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala:638]: For my understanding: Once we fix the V2SessionCatalog bug in Apache Spark, I guess we can always pass a `createTableFunc` in all cases. How do we take care of `ignoreIfExists` and `validateLocation` parameters? i.e. what is the default behavior of `super.createTable(ident, t.columns(), t.partitioning, t.properties)` w.r.t. `ignoreIfExists` and `validateLocation` parameters?

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 4]
```diff
@@ -181,7 +187,21 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
       operation,
-      tableByPath = isByPath).run(spark)
+      tableByPath = isByPath,
+      // We should invoke the Spark catalog plugin API to create the table, to
+      // respect third party catalogs. Note: only handle CREATE TABLE for now, we
+      // should support CTAS later.
+      // TODO: Spark `V2SessionCatalog` mistakenly treat tables with location as EXTERNAL table.
+      //       Before this bug is fixed, we should only call the catalog plugin API to create tables
+      //       if UC is enabled to replace `V2SessionCatalog`.
+      createTableFunc = if (isUnityCatalog && sourceQuery.isEmpty) {
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:197]: Is there a way to add unit test for this change? I guess no as UC side changes needs to be released first?

Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:197]: Yea, it's a UC-specific hack and we do have tests at the UC side.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala (hunk 3 above)
Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala:638]: We will need to catch `TableAlreadyExistsException` to handle `ifExists`. There is no location validation in v2 catalog `createTable` so we are fine.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala (hunk 2 above)
Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala:73]: done

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 5]
```diff
@@ -70,6 +70,12 @@ class DeltaCatalog extends DelegatingCatalogExtension
 
   val spark = SparkSession.active
 
+  private lazy val isUnityCatalog: Boolean = {
+    val delegateField = classOf[DelegatingCatalogExtension].getDeclaredField("delegate")
```
Comment (sumeet-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:74]: Is the `delegate` field guaranteed to exist?

Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:74]: Yes: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/DelegatingCatalogExtension.java#L41

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 6]
```diff
@@ -70,6 +70,12 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
+    delegateField.setAccessible(true)
+    delegateField.get(this).getClass.getCanonicalName.startsWith("io.unitycatalog")
```
Comment (scovich) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:76]:
```suggestion
    delegateField.get(this).getClass.getCanonicalName.startsWith("io.unitycatalog.")
```
... to at least slightly reduce the chances of a false positive?

**Review by prakharjain09 — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 7]
```diff
@@ -181,7 +187,21 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
+      // TODO: Spark `V2SessionCatalog` mistakenly treat tables with location as EXTERNAL table.
+      //       Before this bug is fixed, we should only call the catalog plugin API to create tables
+      //       if UC is enabled to replace `V2SessionCatalog`.
+      createTableFunc = if (isUnityCatalog && sourceQuery.isEmpty) {
+        Some(v1Table => {
+          val t = V1Table(v1Table)
+          super.createTable(ident, t.columns(), t.partitioning, t.properties)
+        })
+      } else {
+        None
```
Comment (scovich) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:203]:
```suggestion
      createTableFunc = Option.when(isUnityCatalog && sourceQuery.isEmpty) { v1table =>
          val t = V1Table(v1Table)
          super.createTable(ident, t.columns(), t.partitioning, t.properties)
```

**Review by scovich — APPROVED** [review]
LGTM, couple nits.

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala (hunk 4 above)
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:197]: Does it need to be specific for unitycatalog?

Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:197]: It has to, until Spark 3.5.3 is released.

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 8]
```diff
@@ -181,7 +187,17 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
+      // We should invoke the Spark catalog plugin API to create the table, to
+      // respect third party catalogs. Note: only handle CREATE TABLE for now, we
+      // should support CTAS later.
+      // TODO: Spark `V2SessionCatalog` mistakenly treat tables with location as EXTERNAL table.
+      //       Before this bug is fixed, we should only call the catalog plugin API to create tables
+      //       if UC is enabled to replace `V2SessionCatalog`.
+      createTableFunc = Option.when(isUnityCatalog && sourceQuery.isEmpty) { v1table =>
+          val t = V1Table(v1Table)
+          super.createTable(ident, t.columns(), t.partitioning, t.properties)
+      }).run(spark)
```
Comment (cloud-fan) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:200]:
```suggestion
      createTableFunc = Option.when(isUnityCatalog && sourceQuery.isEmpty) { v1table =>
        val t = V1Table(v1Table)
        super.createTable(ident, t.columns(), t.partitioning, t.properties)
      }).run(spark)
```

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala (hunk 4 above)
Comment (felipepessoto) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala]: Got it. I’d only suggest adding a TODO comment. Thanks

**Comment by cloud-fan** [comment]
`Option.when` is available after Scala 2.13

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala [hunk 9]
```diff
@@ -181,7 +187,17 @@ class DeltaCatalog extends DelegatingCatalogExtension
...
+      // We should invoke the Spark catalog plugin API to create the table, to
+      // respect third party catalogs. Note: only handle CREATE TABLE for now, we
+      // should support CTAS later.
+      // TODO: Spark `V2SessionCatalog` mistakenly treat tables with location as EXTERNAL table.
+      //       Before this bug is fixed, we should only call the catalog plugin API to create tables
+      //       if UC is enabled to replace `V2SessionCatalog`.
+      createTableFunc = Option.when(isUnityCatalog && sourceQuery.isEmpty) { v1table =>
+        val t = V1Table(v1Table)
+        super.createTable(ident, t.columns(), t.partitioning, t.properties)
+      }).run(spark)
```
Comment (cloud-fan) [inline spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:200]:
```suggestion
      createTableFunc = if (isUnityCatalog && sourceQuery.isEmpty) {
        Some(v1Table => {
          val t = V1Table(v1Table)
          super.createTable(ident, t.columns(), t.partitioning, t.properties)
        })
      } else {
        None
      }).run(spark)
```

File: spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala (hunk 4 above)
Comment (cloud-fan) [reply spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:197]: There is already a TODO for it.

**Comment by scovich** [comment]
> `Option.when` is available after Scala 2.13

It's also available in 2.12 by importing the compat shim in https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/util/ScalaExtensions.scala#L36

**Comment by tdas** [comment]
spark master failures seems un related.

**Comment by cloud-fan** [comment]
The master branch failure seems to be caused by the recent Spark refactors for Spark Connect. cc @hvanhovell

**Comment by tdas** [comment]
There seems to be other spark test failures.

**Comment by tdas** [comment]
@cloud-fan can you fix them?

**Comment by tdas** [comment]
merged. can you make a PR for branch 3.2?

**Comment by cloud-fan** [comment]
the 3.2 PR: https://github.com/delta-io/delta/pull/3603

**Pull request merged**